drivers/base/regmap/regmap-sdw-mbq.c

```diff
@@ -36,30 +36,38 @@ static int regmap_sdw_mbq_write(void *context, unsigned int reg, unsigned int va
 	struct device *dev = ctx->dev;
 	struct sdw_slave *slave = dev_to_sdw_dev(dev);
 	int mbq_size = regmap_sdw_mbq_size(ctx, reg);
-	int ret;
-
-	switch (mbq_size) {
-	case 4:
-		ret = sdw_write_no_pm(slave, SDW_SDCA_MBQ_CTL(reg),
-				      (val >> 24) & 0xff);
-		if (ret < 0)
-			return ret;
-
-		ret = sdw_write_no_pm(slave, SDW_SDCA_MBQ_CTL(reg),
-				      (val >> 16) & 0xff);
-		if (ret < 0)
-			return ret;
-		fallthrough;
-	case 2:
-		ret = sdw_write_no_pm(slave, SDW_SDCA_MBQ_CTL(reg),
-				      (val >> 8) & 0xff);
-		if (ret < 0)
-			return ret;
-		fallthrough;
-	case 1:
-		return sdw_write_no_pm(slave, reg, val & 0xff);
-	default:
-		return -EINVAL;
+	int i, ret;
+
+	for (i = 0; i < ctx->cfg.retry_n; i++) {
+		switch (mbq_size) {
+		case 4:
+			ret = sdw_write_no_pm(slave, SDW_SDCA_MBQ_CTL(reg),
+					      (val >> 24) & 0xff);
+			if (ret < 0)
+				return ret;
+
+			ret = sdw_write_no_pm(slave, SDW_SDCA_MBQ_CTL(reg),
+					      (val >> 16) & 0xff);
+			if (ret < 0)
+				return ret;
+			fallthrough;
+		case 2:
+			ret = sdw_write_no_pm(slave, SDW_SDCA_MBQ_CTL(reg),
+					      (val >> 8) & 0xff);
+			if (ret < 0)
+				return ret;
+			fallthrough;
+		case 1:
+			ret = sdw_write_no_pm(slave, reg, val & 0xff);
+			if (ret != -ENODATA)
+				return ret;
+
+			dev_dbg(dev, "Retry write to register 0x%x\n", reg);
+			usleep_range(ctx->cfg.retry_us, ctx->cfg.retry_us + 100);
+			break;
+		default:
+			return -EINVAL;
+		}
```

Comment:
I have multiple objections to this patch

a) what makes you think the COMMAND_IGNORED will only happen for MBQs?

b) there's no concept of retry time, only a worst-case property, and the best solution is really to poll on the FunctionBusy bit. even if you only use the sleep and retry, for the final timeout case you'd want to log what the function busy bit value is.

c) DisCo has the concept of mipi-sdca-control-deferrable, so we should only use the deferred mechanism for the controls (and thus address ranges) that require this mechanism.

I would use another callback in addition to mbq_size to check if a control requires this check or not.

Reply:
Using the function busy would perhaps be slightly more accurate, but it adds a fair amount of complexity without really much in the way of advantages. You still have all the same questions of how often to poll and how long to poll for, I wouldn't object to changing to a total poll time if that is preferred rather than a delay between polls but fairly close to equivalent.

I could perhaps add a callback to trigger this only for deferrable controls, the decision I made the first time through was that it wasn't really worth it. There are really two situations you could get a command ignored for a non-deferrable control, the control doesn't exist or something went wrong. In either case retrying didn't really feel like much of an issue. It will either fail eventually or succeed on the retry if it was an intermittent fault.

Reply:
if you have an intermittent fault then something is really very wrong -> the device lost sync, interrupts are disabled and the audio transfers were stopped. I don't see much merit in trying to conceal such problems, fail big and fail early is better IMHO. I really think we should only have a different behavior when the spec requires us to do so...

Reply:
Yeah that is a fair point, I do like to fail early and big :-) will add a callback for the deferrabililty of the control.